Event delegation (#203)
javascript-tutorial/it.javascript.info: 2-ui/2-events/03-event-delegation/article.md

# Event delegation

Capturing e bubbling ci permettono di implementare uno dei pattern più potenti nella gestione degli eventi, e cioè *event delegation*.

> Il concetto alla base è che se abbiamo una serie di elementi gestiti in maniera simile, allora invece di assegnare un gestore per ognuno di essi, possiamo metterne uno solo sui loro antenati comuni.

Reviewer: ```suggestion
Il concetto di base è che se abbiamo una serie di elementi gestiti in maniera simile, allora, invece di assegnare un gestore per ognuno di essi, possiamo metterne uno solo sui loro antenati comuni.
```

Author: /done

> Nel gestore avremo a disposizione `event.target` per controllare l'elemento dal quale sia partito l'evento e poterlo quindi gestire di conseguenza.

Reviewer 2: ```suggestion
Nel gestore avremo a disposizione `event.target` per controllare l'elemento dal quale è partito l'evento e poterlo quindi gestire di conseguenza.
```

Author: /done

</table>
```

La tabella è composta da 9 celle, ma potrebbero anche essercene 99 o 9999, è irrilevante.

**Il nostro compito è quello di evidenziare un cella `<td>` al click.**

> Invece di assegnare un gestore all'`onclick` per ogni `<td>` (potrebbero essercene tantissime) -- andiamo a impostare un gestore del tipo "catturali tutti" sull'elemento `<table>`.

Reviewer 2: non mi convince tantissimo questa sezione, piccolo suggerimento
```suggestion
Invece di assegnare un gestore all'`onclick` per ogni `<td>` (potrebbero essercene tantissime) -- andiamo a impostare un gestore che sarà in grado di "catturali tutti" sull'elemento `<table>`.
```

Author: no, ma ok.

Author: /done

};
```

Chiarimenti:
1. Il metodo `elem.closest(selector)` ritorna l'antenato più vicino che combacia con il selettore. Nel nostro caso cerchiamo un `<td>` verso l'alto dall'elemento di origine dell'evento.
2. Se `event.target` non è dentro nessun `<td>`, la chiamata esce immediatamente, dal momento che non c'è nulla da fare.

> 3. Ne caso di tabelle annidate, `event.target` potrebbe essere un `<td>`, ma fuori dalla tabelle corrente. Quindi andiamo a controllare se sia il `<td>` della *nostra tabella*.

Reviewer 2: ```suggestion
3. Nel caso di tabelle annidate, `event.target` potrebbe essere un `<td>`, ma fuori dalla tabelle corrente. Quindi andiamo a controllare se sia il `<td>` della *nostra tabella*.
```

Author: /done

> 3. Ne caso di tabelle annidate, `event.target` potrebbe essere un `<td>`, ma fuori dalla tabelle corrente. Quindi andiamo a controllare se sia il `<td>` della *nostra tabella*.

Reviewer 2: ```suggestion
3. Ne caso di tabelle annidate, `event.target` potrebbe riferirsi ad `<td>`, ma fuori dalla tabelle corrente. Quindi andiamo a controllare se `<td>`  appartiene alla *nostra tabella*.
```

4. E se così, la evidenziamo.

> Come risultato, averemo un codice di evidenziazione veloce ede efficiente, indipendente dal numero di `<td>` nella tabella.

Author: /done

## Esempio di delegation: azioni nel markup

Esistono altri utilizzi per l'event delegation.

> Poniamo il caso che volessimo fare un menù con i pulsanti "Save", "Load", "Search" e cosi via, e che via sia un oggetto con i metodi `save`, `load`, `search`... Come potremmo distinguerli?

Reviewer: ```suggestion
Poniamo il caso che volessimo fare un menù con i pulsanti "Save", "Load", "Search" e cosi via, e che vi sia un oggetto con i metodi `save`, `load`, `search`... Come potremmo distinguerli?
```

Author: /done

</script>
```

> Nota bene che `this.onClick` è collegato a `this` nel punto `(*)`.  Questo è importante, altrimenti `this` potrebbe riferirsi all'elemento del DOM (`elem`), e non l'oggetto `Menu`, di conseguenza `this[action]` non farebbe quello di cui abbiamo bisogno.

Reviewer: ```suggestion
Nota bene che `this.onClick` è collegato a `this` nel punto `(*)`.  Questo è importante, altrimenti `this` si riferirebbe all'elemento del DOM (`elem`), e non l'oggetto `Menu`, di conseguenza `this[action]` non farebbe quello di cui abbiamo bisogno.
```

Author: /done

Quindi, quali vantaggi apporta delegation qui?

```compare
+ Non abbiamo bisogno di scrivere del codice per assegnare un gestore ad ogni pulsante. Ma solo di un metodo e porlo dentro il markup.
+ La struttura HTML è flessibile, possiamo aggiungere e rimuovere pulsanti in ogni momento.
```

Possiamo anche usare classi come `.action-save`, `.action-load`, ma un attributo `data-action` è semanticamente migliore. Inoltre possiamo usarlo nelle regole CSS.

## Il pattern "comportamentale"

Possiamo anche usare la event delegation per aggiungere "comportamenti" agli elementi in modo *dichiarativo*, con speciali attributi e classi.

Il pattern consta di due parti:
1. Aggiungiamo un attributo personalizzato a un elemento, che descrive il suo comportamento.

> 2. Un gestore su tutto il documento tiene traccia degli eventi, e se viene scaturito un evento su un elemento con l'attributo -- esegue l'azione.

Reviewer: ```suggestion
2. Un gestore su tutto il documento tiene traccia degli eventi, e se viene attivato un evento su un elemento con quell'attributo -- esegue l'azione.
```

Author: /done

> ### Comportamento: contator

Reviewer 2: ```suggestion
### Comportamento: contatore
```

> Per esempio, qui l'attributo `data-counter` aggiunge un comportamento: "aumenta il valore al click" sui pulsanti:

Reviewer: ```suggestion
Per esempio, qui l'attributo `data-counter` aggiunge un comportamento: "incrementa il valore al click" sui pulsanti:
```

Author: /done

```html run autorun height=60
Counter: <input type="button" value="1" data-counter>
Contatore di addizione: <input type="button" value="2" data-counter>

<script>
  document.addEventListener('click', function(event) {

    if (event.target.dataset.counter != undefined) { // se esiste l'attributo...
      event.target.value++;
    }

  });
</script>
```

Se clicchiamo un pulsante -- il suo valore aumenterà. Non sono importanti i pulsanti qui, ma l'approccio in generale.

Possono esserci quanti attributi `data-counter` vogliamo. Possiamo aggiungerne di nuovi all'HTML in ogni momento. Usando la event delegation abbiamo "esteso" HTML, aggiunto un attributo che descrive un nuovo comportamento.

> ```warn header="Pe gestori a livello di documento -- usare sempre `addEventListener`"

Reviewer: ```suggestion
```warn header="Per gestori a livello di documento -- usare sempre `addEventListener`"
```

Author: /done

> Quando assegniamo un gestore di eventi all'oggetto `document`, dovremmo sempre usare `addEventListener`, e non `document.on<event>`, perché il secondo causerebbe conflitti: dei nuovi gestori sovrascriverebbero i precedenti.

Reviewer: ```suggestion
Quando assegniamo un gestore di eventi all'oggetto `document`, dovremmo sempre usare `addEventListener`, e non `document.on<event>`, perché il secondo causerebbe conflitti: i nuovi gestori sovrascriverebbero i precedenti.
```

Author: /done

</script>
```

Notiamo ancora una volta cosa abbiamo fatto. Adesso, per aggiungere la funzionalità di toggling su un elemento -- non è necessario conoscere JavaScript, è sufficiente usare l'attributo `data-toggle-id`.

> Questo può essere davvero conveniente -- nessuna necessità di scrivere codice JavaScript per ogni nuovo elemento di questo genere. Basta solo usare il comportamento. Il gestore a livello di documento fa in modo che funzioni per ogni elemento nella pagina.

Reviewer 2: ```suggestion
Questo può essere davvero conveniente -- nessuna necessità di scrivere codice JavaScript per ogni nuovo elemento di questo genere. Ci basta solo applicare il comportamento. Il gestore a livello di documento fa in modo che funzioni per ogni elemento nella pagina.
```
